### ISSUE TITLE
Setup fails because the file is locked

### ISSUE THREAD

**Issue by KirillOsenkov** [NickeManarin/ScreenToGif#876]
I keep my ScreenToGif.exe in my Dropbox folder, and when I updated today the setup failed with this message in the log:

```
▬ Message - 
	The process cannot access the file 'C:\Dropbox\Tools\ProgramFilesFolder\ScreenToGif' because it is being used by another process.
♦ Date/Hour - 
	5/2/2021 8:14:35 PM
▲ Source - 
	mscorlib
▼ TargetSite - 
	Void WinIOError(Int32, System.String)
♠ StackTrace - 
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.Directory.DeleteHelper(String fullPath, String userPath, Boolean recursive, Boolean throwOnTopLevelDirectoryNotFound, WIN32_FIND_DATA& data)
   at System.IO.Directory.Delete(String fullPath, String userPath, Boolean recursive, Boolean checkHost)
   at ScreenToGif_Installer.UserControls.Progress.CopyPortableFiles()
```

I'm guessing Dropbox takes some kind of a lock on files intermittently so perhaps consider adding a retry a few times?

To help with diagnosing locked files, consider using https://github.com/cklutz/LockCheck (here's a sample of how it's used in MSBuild: https://source.dot.net/#Microsoft.Build.Tasks.Core/LockCheck.cs,726a666e72008c9b)

Thanks again for the fantastic tool!

**Comment by NickeManarin** [comment]
Looks like that you are using the portable variant, is that right?

The update process calls the installer to run a custom action.
It looks like doable.

Thanks for the feedback and the proposed solution.

**Comment by NickeManarin** [comment]
The installer will try to copy/remove files multiple (~4).

**Comment by KirillOsenkov** [comment]
Hmm, I've tried to update to 2.30.1 and got Setup failed again.

Here's the log:
https://gist.github.com/KirillOsenkov/41d202ebb71cab3ed31c6ef8d28b4413

For some reason it tries to access the folder `D:\Dropbox\MS\Tools\ProgramFilesFolder\ScreenToGif`, which I don't remember creating.

Now it has two files:
![image](https://user-images.githubusercontent.com/679326/119766957-d8521180-be6a-11eb-8bd9-e53cc1778b4b.png)

I'd saved ScreenToGif.exe to `D:\Dropbox\MS\Tools\ScreenToGif.exe` and that's where I'm running it from.

But later I did try run ScreenToGif.exe and it started the version 2.30.1 so it must have updated successfully (mostly).

**Comment by KirillOsenkov** [comment]
Consider also adding more logging, like which process is running (`Process.GetCurrentProcess().MainModule.FileName`), etc.

**Comment by KirillOsenkov** [comment]
![image](https://user-images.githubusercontent.com/679326/119767189-41398980-be6b-11eb-9e58-42a5714f05d0.png)

**Comment by NickeManarin** [comment]
Ohh, Ok. 

I'm going to change my approach. 
Thank you for the log.

The installer in `/a` mode creates that folder when unpacking the files.
I just erase it after getting the necessary files from it.

**Comment by NickeManarin** [comment]
I changed the process, now it waits indefinitely until the process releases the file. 

![Update](https://user-images.githubusercontent.com/14798947/120087639-0bd9a980-c0c0-11eb-94af-a29983e5461f.gif)

I'm using some simple Ntdll calls.

**Comment by KirillOsenkov** [comment]
Thanks! Out of curiosity, where's the logic that checks for blocking files? I couldn't find it in the repo. Curious which ntdll calls you're using.

**Comment by NickeManarin** [comment]
It's closed source. 
I'm using something similar to LockCheck (but way more simple).
It's similar enough that I should add a license info. (I just added now, it will be available with v2.32)

**Comment by KirillOsenkov** [comment]
I'm just intrigued because I don't know of any other solution other than LockCheck, and I don't see how I would simplify it. If you can't share, that's OK, but if you can post a sample I'd be very curious to learn about it.
